File: src/main/java/jaxb/JsonToJaxbImpl.java
package jaxb;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import javax.xml.XMLConstants;
import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBElement;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Marshaller;
import javax.xml.namespace.QName;
import java.io.StringWriter;
import java.lang.reflect.Type;


public class JsonToJaxbImpl implements JsonToJaxb {

    private static final String UTF_8 = "UTF-8";
    private Gson gson;

    public static JsonToJaxbImpl create(GsonBuilder builder) {
        return new JsonToJaxbImpl(builder.create());
    }

    public static JsonToJaxbImpl create() {
        return new JsonToJaxbImpl(
                new GsonBuilder().create()
        );
    }

    JsonToJaxbImpl(Gson gson){
        this.gson = gson;
    }

    public <T> T toClass(String json, Type _type){
        final T newclass = gson.fromJson(json, _type);
        return newclass;
    }

    @SuppressWarnings("static-access")
    public <T> T toClass(String json){
        final Type type =  this.<T>factoryTypeToken();
        final T newclass = gson.fromJson(json, type);
        return newclass;
    }

    public <T> String toJson(T clazz, Type _type){
        final String toJson = gson.toJson(clazz,_type);
        return toJson;
    }

    public <T> String toJson(T clazz){
        final Type type =  factoryTypeToken();
        final String toJson = toJson(clazz,type);
        return toJson;
    }

    public <T> String ToXml(T obj) throws JAXBException{
        final JAXBContext jaxbContext = JAXBContext.newInstance(obj.getClass());
        final Marshaller jaxbMarshaller = jaxbContext.createMarshaller();

        final StringWriter sw = new StringWriter();

        jaxbMarshaller.setProperty(Marshaller.JAXB_ENCODING, UTF_8);
        jaxbMarshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, true);

        jaxbMarshaller.marshal(obj, sw);

        return sw.toString();
    }

    public <I,O> String ToXml(String json, Type _type) throws JAXBException{
        final I obj = toClass(json,_type);
        String res = ToXml(obj);

        return res;
    }

    public <T> String ToJaxbElementString(T obj) throws JAXBException{
        return ToJaxbElementString(XMLConstants.NULL_NS_URI,"",XMLConstants.DEFAULT_NS_PREFIX,obj);
    }

    public <T> String ToJaxbElementString(String local, T obj) throws JAXBException{
        return ToJaxbElementString(XMLConstants.NULL_NS_URI,local,XMLConstants.DEFAULT_NS_PREFIX,obj);
    }

    public <T> String ToJaxbElementString(String uri, String local, T obj) throws JAXBException{
        return ToJaxbElementString(uri,local,XMLConstants.DEFAULT_NS_PREFIX,obj);
    }

    public <T> String ToJaxbElementString(String uri, String local, String prefix, T obj) throws JAXBException{
        final JAXBContext jaxbContext = JAXBContext.newInstance(obj.getClass());
        final Marshaller jaxbMarshaller = jaxbContext.createMarshaller();
        final JAXBElement<T> element = ToJaxbElement(uri,local,prefix, obj);
        jaxbMarshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, Boolean.TRUE);

        final StringWriter sw = new StringWriter();
        jaxbMarshaller.marshal(element, sw);

        return sw.toString();

    }

    public <T> JAXBElement<T> ToJaxbElement(T obj) throws JAXBException{
        final JAXBElement<T> element = new JAXBElement<T>(new QName(XMLConstants.NULL_NS_URI,"",XMLConstants.DEFAULT_NS_PREFIX),  getClass(obj), obj);
        return element;
    }

    public <T> JAXBElement<T> ToJaxbElement(String local, T obj) throws JAXBException{
        final JAXBElement<T> element = new JAXBElement<T>(new QName(XMLConstants.NULL_NS_URI,local,XMLConstants.DEFAULT_NS_PREFIX),  getClass(obj), obj);
        return element;
    }


    public <T> JAXBElement<T> ToJaxbElement(String uri, String local, T obj) throws JAXBException{
        final JAXBElement<T> element = new JAXBElement<T>(new QName(uri,local,XMLConstants.DEFAULT_NS_PREFIX),  getClass(obj), obj);
        return element;
    }

    public <T> JAXBElement<T> ToJaxbElement(String uri, String local, String prefix, T obj) throws JAXBException{
        final JAXBElement<T> element = new JAXBElement<T>(new QName(uri,local,prefix),   getClass(obj), obj);
        return element;
    }

    public <T> JAXBElement<T> ToJaxbElement(String fullPathNamespaceClass, String json, Type type) throws JAXBException, ClassNotFoundException{
        T objJson = this.<T>getClassWithFullNameObject(fullPathNamespaceClass, json,type);
        final JAXBElement<T> element = ToJaxbElement(objJson);
        return element;
    }

    public <T> JAXBElement<T> ToJaxbElement(String fullPathNamespaceClass, String json) throws JAXBException, ClassNotFoundException, InstantiationException, IllegalAccessException{
        T objJson = this.<T>getClassWithFullNameObject(fullPathNamespaceClass, json);
        final JAXBElement<T> element = ToJaxbElement(objJson);
        return element;
    }

    public <T> JAXBElement<T> ToJaxbElement(String local, String fullPathNamespaceClass, String json, Type type) throws JAXBException, ClassNotFoundException{
        T objJson = this.<T>getClassWithFullNameObject(fullPathNamespaceClass, json,type);
        final JAXBElement<T> element = ToJaxbElement(local, objJson);
        return element;
    }

    public <T> JAXBElement<T> ToJaxbElement(String uri, String local, String fullPathNamespaceClass, String json, Type type) throws JAXBException, ClassNotFoundException{
        T objJson = this.<T>getClassWithFullNameObject(fullPathNamespaceClass, json,type);
        final JAXBElement<T> element = ToJaxbElement(uri,local, objJson);
        return element;
    }

    public <T> JAXBElement<T> ToJaxbElement(String uri, String local, String prefix, String fullPathNamespaceClass, String json, Type type) throws JAXBException, ClassNotFoundException{
        T objJson = this.<T>getClassWithFullNameObject(fullPathNamespaceClass, json, type);
        final JAXBElement<T> element = ToJaxbElement(uri,local,prefix, objJson);
        return element;
    }

    @SuppressWarnings("static-access")
    public <T> JAXBElement<T> ToJaxbElement(String local, String fullPathNamespaceClass, String json) throws JAXBException, ClassNotFoundException{
        T objJson = this.<T>getClassWithFullNameObject(fullPathNamespaceClass, json);
        final JAXBElement<T> element = ToJaxbElement(local, objJson);
        return element;
    }

    @SuppressWarnings("static-access")
    public <T> JAXBElement<T> ToJaxbElement(String uri, String local, String fullPathNamespaceClass, String json) throws JAXBException, ClassNotFoundException{
        T objJson = this.<T>getClassWithFullNameObject(fullPathNamespaceClass, json );
        final JAXBElement<T> element = ToJaxbElement(uri,local, objJson);
        return element;
    }

    public <T> JAXBElement<T> ToJaxbElement(String uri, String local, String prefix, String fullPathNamespaceClass, String json) throws JAXBException, ClassNotFoundException{
        T objJson = this.<T>getClassWithFullNameObject(fullPathNamespaceClass, json);
        final JAXBElement<T> element = ToJaxbElement(uri,local,prefix, objJson);
        return element;
    }

    private static <T> Type factoryTypeToken() {
        return new TypeToken<T>(){}.getType();
    }

    @SuppressWarnings({ "rawtypes"})
    private <T> T getClassWithFullNameObject(String fullPathNamespaceClass, String json) throws ClassNotFoundException {
        Class obj =  Class.forName(fullPathNamespaceClass);
        T objJson = (T) gson.fromJson(json, obj);
        return objJson;
    }

    @SuppressWarnings({ "rawtypes", "unused" })
    private <T> T getClassWithFullNameObject(String fullPathNamespaceClass, String json, Type type) throws ClassNotFoundException {
        Class obj =  Class.forName(fullPathNamespaceClass);
        T objJson = gson.fromJson(json, type);
        return objJson;
    }

    @SuppressWarnings("unchecked")
    private <T> Class<T> getClass(T obj) {
        return (Class<T>) obj.getClass();
    }

}

File: src/main/java/jaxb/JsonToJaxbImpl.kt
package jaxb

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import java.io.StringWriter
import java.lang.reflect.Type
import javax.xml.XMLConstants
import javax.xml.bind.JAXBContext
import javax.xml.bind.JAXBElement
import javax.xml.bind.JAXBException
import javax.xml.bind.Marshaller
import javax.xml.namespace.QName
import kotlin.reflect.KClass


class JsonToJaxbImpl internal constructor(private val gson: Gson) : JsonToJaxb {

    @SuppressWarnings("unchecked")
    private fun<T: Any> T.getClass(): KClass<T> = javaClass.kotlin

    companion object {

        private val UTF_8 = "UTF-8"

        fun create(builder: GsonBuilder) = JsonToJaxbImpl(builder.create())

        fun create(): JsonToJaxbImpl = JsonToJaxbImpl(GsonBuilder().create())

        private fun <T> factoryTypeToken() = object : TypeToken<T>() {}.type
    }

    override fun <T : Any> toClass(json: String, _type: Type) = gson.fromJson<T>(json, _type)

    @SuppressWarnings("static-access")
    override fun <T : Any> toClass(json: String): T {
        val type = factoryTypeToken<T>()
        val newclass = gson.fromJson<T>(json, type)
        return newclass
    }

    override fun <T : Any> toJson(clazz: T, _type: Type) =  gson.toJson(clazz, _type)

    override fun <T : Any> toJson(clazz: T) = toJson(clazz, factoryTypeToken<T>())

    @Throws(JAXBException::class)
    override fun <T : Any> ToXml(obj: T): String {
        val jaxbContext = JAXBContext.newInstance(obj.getClass().java)
        val jaxbMarshaller = jaxbContext.createMarshaller()

        val sw = StringWriter()

        jaxbMarshaller.setProperty(Marshaller.JAXB_ENCODING, UTF_8)
        jaxbMarshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, true)

        jaxbMarshaller.marshal(obj, sw)

        return sw.toString()
    }

    @Throws(JAXBException::class)
    override fun <I : Any, O> ToXml(json: String, _type: Type): String {
        val obj = toClass<I>(json, _type)
        val res = ToXml(obj)

        return res
    }

    @Throws(JAXBException::class)
    override fun <T : Any> ToJaxbElementString(obj: T): String {
        return ToJaxbElementString(XMLConstants.NULL_NS_URI, "", XMLConstants.DEFAULT_NS_PREFIX, obj)
    }

    @Throws(JAXBException::class)
    override fun <T : Any> ToJaxbElementString(local: String, obj: T): String {
        return ToJaxbElementString(XMLConstants.NULL_NS_URI, local, XMLConstants.DEFAULT_NS_PREFIX, obj)
    }

    @Throws(JAXBException::class)
    override fun <T : Any> ToJaxbElementString(uri: String, local: String, obj: T): String {
        return ToJaxbElementString(uri, local, XMLConstants.DEFAULT_NS_PREFIX, obj)
    }

    @Throws(JAXBException::class)
    override fun <T : Any> ToJaxbElementString(uri: String, local: String, prefix: String, obj: T): String {
        val jaxbContext = JAXBContext.newInstance(obj.getClass().java)
        val jaxbMarshaller = jaxbContext.createMarshaller()
        val element = ToJaxbElement(uri, local, prefix, obj)
        jaxbMarshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, java.lang.Boolean.TRUE)

        val sw = StringWriter()
        jaxbMarshaller.marshal(element, sw)

        return sw.toString()

    }

    @Throws(JAXBException::class)
    override fun <T : Any> ToJaxbElement(obj: T): JAXBElement<T> {
        val element = JAXBElement(QName(XMLConstants.NULL_NS_URI, "", XMLConstants.DEFAULT_NS_PREFIX), obj.getClass().java, obj)
        return element
    }

    @Throws(JAXBException::class)
    override fun <T : Any> ToJaxbElement(local: String, obj: T): JAXBElement<T> {
        val element = JAXBElement(QName(XMLConstants.NULL_NS_URI, local, XMLConstants.DEFAULT_NS_PREFIX), obj.getClass().java, obj)
        return element
    }


    @Throws(JAXBException::class)
    override fun <T : Any> ToJaxbElement(uri: String, local: String, obj: T): JAXBElement<T> {
        val element = JAXBElement(QName(uri, local, XMLConstants.DEFAULT_NS_PREFIX), obj.getClass().java, obj)
        return element
    }

    @Throws(JAXBException::class)
    override fun <T : Any> ToJaxbElement(uri: String, local: String, prefix: String, obj: T): JAXBElement<T> {
        val element = JAXBElement(QName(uri, local, prefix), obj.getClass().java, obj)
        return element
    }

    @Throws(JAXBException::class, ClassNotFoundException::class)
    override fun <T : Any> ToJaxbElement(fullPathNamespaceClass: String, json: String, type: Type): JAXBElement<T> {
        val objJson = this.getClassWithFullNameObject<T>(fullPathNamespaceClass, json, type)
        val element = ToJaxbElement(objJson)
        return element
    }

    @Throws(JAXBException::class, ClassNotFoundException::class, InstantiationException::class, IllegalAccessException::class)
    override fun <T : Any> ToJaxbElement(fullPathNamespaceClass: String, json: String): JAXBElement<T> {
        val objJson = this.getClassWithFullNameObject<T>(fullPathNamespaceClass, json)
        val element = ToJaxbElement(objJson)
        return element
    }

    @Throws(JAXBException::class, ClassNotFoundException::class)
    override fun <T : Any> ToJaxbElement(local: String, fullPathNamespaceClass: String, json: String, type: Type): JAXBElement<T> {
        val objJson = this.getClassWithFullNameObject<T>(fullPathNamespaceClass, json, type)
        val element = ToJaxbElement(local, objJson)
        return element
    }

    @Throws(JAXBException::class, ClassNotFoundException::class)
    override fun <T : Any> ToJaxbElement(uri: String, local: String, fullPathNamespaceClass: String, json: String, type: Type): JAXBElement<T> {
        val objJson = this.getClassWithFullNameObject<T>(fullPathNamespaceClass, json, type)
        val element = ToJaxbElement(uri, local, objJson)
        return element
    }

    @Throws(JAXBException::class, ClassNotFoundException::class)
    override fun <T : Any> ToJaxbElement(uri: String, local: String, prefix: String, fullPathNamespaceClass: String, json: String, type: Type): JAXBElement<T> {
        val objJson = this.getClassWithFullNameObject<T>(fullPathNamespaceClass, json, type)
        val element = ToJaxbElement(uri, local, prefix, objJson)
        return element
    }

    @SuppressWarnings("static-access")
    @Throws(JAXBException::class, ClassNotFoundException::class)
    override fun <T : Any> ToJaxbElement(local: String, fullPathNamespaceClass: String, json: String): JAXBElement<T> {
        val objJson = this.getClassWithFullNameObject<T>(fullPathNamespaceClass, json)
        val element = ToJaxbElement(local, objJson)
        return element
    }

    @SuppressWarnings("static-access")
    @Throws(JAXBException::class, ClassNotFoundException::class)
    override fun <T : Any> ToJaxbElement(uri: String, local: String, fullPathNamespaceClass: String, json: String): JAXBElement<T> {
        val objJson = this.getClassWithFullNameObject<T>(fullPathNamespaceClass, json)
        val element = ToJaxbElement(uri, local, objJson)
        return element
    }

    @Throws(JAXBException::class, ClassNotFoundException::class)
    override fun <T : Any> ToJaxbElement(uri: String, local: String, prefix: String, fullPathNamespaceClass: String, json: String): JAXBElement<T> {
        val objJson = this.getClassWithFullNameObject<T>(fullPathNamespaceClass, json)
        val element = ToJaxbElement(uri, local, prefix, objJson)
        return element
    }

    @SuppressWarnings("rawtypes")
    @Throws(ClassNotFoundException::class)
    private fun <T> getClassWithFullNameObject(fullPathNamespaceClass: String, json: String): T {
        val obj = Class.forName(fullPathNamespaceClass)
        val objJson = gson.fromJson(json, obj) as T
        return objJson
    }

    @SuppressWarnings("rawtypes", "unused")
    @Throws(ClassNotFoundException::class)
    private fun <T> getClassWithFullNameObject(fullPathNamespaceClass: String, json: String, type: Type): T {
        val obj = Class.forName(fullPathNamespaceClass)
        val objJson = gson.fromJson<T>(json, type)
        return objJson
    }
}


